pkg/ccl/changefeedccl/topic.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package changefeedccl
+
+import (
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/errors"
+)
+
+// TopicDescriptor describes topic emitted by the sink.
+type TopicDescriptor interface {
+	// GetNameComponents returns the names that should go into any string-based topic identifier.
+	// If the topic is a table, this will be the statement time name of the table, fully-qualified
+	// if that option was set when creating the changefeed.
+	GetNameComponents() []string
+	// GetTopicIdentifier returns a struct suitable for use as a unique key in a map containing
+	// all topics in a feed.
+	GetTopicIdentifier() TopicIdentifier
+	// GetVersion returns topic version.
+	// For example, the underlying data source (e.g. table) may change, in which case
+	// we may want to emit same Name/ID, but a different version number.
+	GetVersion() descpb.DescriptorVersion
+	// GetTargetSpecification() returns the target specification for this topic.
+	// Currently this is assumed to be 1:1.
+	GetTargetSpecification() jobspb.ChangefeedTargetSpecification
+}
+
+// TopicIdentifier is a minimal set of fields that
+// uniquely identifies a topic.
+type TopicIdentifier struct {
+	TableID  descpb.ID
+	FamilyID descpb.FamilyID
+}
+
+// TopicNamer generates and caches the strings used as topic keys by sinks,
+// using target specifications, options, and sink-specific string manipulation.
+type TopicNamer struct {
+	join       byte
+	prefix     string
+	singleName string
+	sanitize   func(string) string
+
+	// DisplayNames are initialized once from specs and may contain placeholder strings.
+	DisplayNames map[jobspb.ChangefeedTargetSpecification]string
+
+	// FullNames are generated whenever Name() is actually called (usually during sink.EmitRow).
+	// They do not contain placeholder strings.
+	FullNames map[TopicIdentifier]string
+
+	sliceCache []string
+}
+
+// MakeTopicNamer creates a TopicNamer.
+// specs are used to populate DisplayNames and the values iterated over in Each.
+// join is the separator for multi-part names, usually '.'.
+// prefix will be prepended to all names.
+// singleName if nonempty will override the normal logic and just return itself no
+// matter what, although prefix and sanitize options will still be honored.
+// sanitize if non-nil will be called to post-process the name.
+func MakeTopicNamer(
+	specs []jobspb.ChangefeedTargetSpecification,
+	join byte,
+	prefix string,
+	singleName string,
+	sanitize func(string) string,
```

Comment:
not particularly thrilled with this API since you always need to add in place `/* .. */` comments when constructing namer.

Reply:
I know this probably involves more typing.  And it just might look like an overkill.
I do like the Options approach (e.g. see `config.go` in `pkg/kv/kvclient/rangefeed`).
I just think that `WithOptionXXX` makes it clear at the call site, what it is you're doing, without having to use these inplace comments (which I happen to dislike).

What would you say if the interface looked like (just an example):
```
func MakeTopicNamer(specs []jobspb.ChangefeedTargetSpecification, opts ...Option) {....}
```

Some of the options could be e.g. `WithJoinRune`, `WithSanitizeFn`, ...

So, I guess at the call site, it would look like:
```
 namer1 := MakeTopicNamer(specs, WithSingleName("foo"))
 namer2 := MakeTopicNamer(specs, WithJoinRune('-'), WithPrefix("bar"), WithSanitizeFn(....))
```